File: pending/2.py
# 2. Add Two Numbers
# https://leetcode.com/problems/add-two-numbers/description/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode() # head: dummy node
        tail = head
        result = 0
        up = 0
        while l1 or l2 or up:
            result = up
            if l1:
                result += l1.val
                l1 = l1.next
            if l2:
                result += l2.val
                l2 = l2.next
            up = result // 10
            tail.next = ListNode(result % 10)
            tail = tail.next
        return head.next

                
        

File: algorithms/swift/2.swift
// 2. Add Two Numbers
// https://leetcode.com/problems/add-two-numbers/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var head = ListNode()
        var node = head

        var l1 = l1
        var l2 = l2
        var extra: Int = 0

        while (l1 != nil) || (l2 != nil) || (extra != 0) {
            let x: Int = l1?.val ?? 0
            let y: Int = l2?.val ?? 0
            l1 = l1?.next
            l2 = l2?.next

            let result: Int = x + y + extra
            extra = result / 10

            node.next = ListNode(result % 10)
            node = node.next!

        }
        return head.next
    }
}
